spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java

```diff
@@ -0,0 +1,395 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.spark.functions;
+
+import java.math.BigInteger;
+import java.nio.ByteBuffer;
+import java.nio.CharBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.util.ByteBuffers;
+import org.apache.iceberg.util.TruncateUtil;
+import org.apache.iceberg.util.UnicodeUtil;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.connector.catalog.functions.BoundFunction;
+import org.apache.spark.sql.connector.catalog.functions.ScalarFunction;
+import org.apache.spark.sql.connector.catalog.functions.UnboundFunction;
+import org.apache.spark.sql.types.BinaryType;
+import org.apache.spark.sql.types.ByteType;
+import org.apache.spark.sql.types.CharType;
+import org.apache.spark.sql.types.DataType;
+import org.apache.spark.sql.types.DataTypes;
+import org.apache.spark.sql.types.Decimal;
+import org.apache.spark.sql.types.DecimalType;
+import org.apache.spark.sql.types.IntegerType;
+import org.apache.spark.sql.types.LongType;
+import org.apache.spark.sql.types.ShortType;
+import org.apache.spark.sql.types.StringType;
+import org.apache.spark.sql.types.StructField;
+import org.apache.spark.sql.types.StructType;
+import org.apache.spark.sql.types.VarcharType;
+import org.apache.spark.unsafe.types.UTF8String;
+
+/**
+ * A function for use in SQL that truncates input to a given width according to the rules of the
+ * existing Truncate partition transform, e.g. {@code SELECT system.truncate(1, 'abc')} will return
+ * the String 'a'.
+ *
+ * <p>Note that for performance reasons, the given input width is not validated in the
+ * implementations used in code-gen. The width must remain non-negative to give meaningful results.
+ */
+public class TruncateFunction implements UnboundFunction {
+  private static final List<DataType> truncateableAtomicTypes =
+      ImmutableList.of(
+          DataTypes.ByteType,
+          DataTypes.ShortType,
+          DataTypes.IntegerType,
+          DataTypes.LongType,
```

Comment:
What about `DecimalType`?

Reply:
This is just the atomic types, as `DataTypes.DecimalType` cannot be used (a precision and scale is needed).

In the validation for the truncation field, we include DecimalType the following way.

```java
  private static void validateTruncationFieldType(DataType dt) {
    if (truncateableAtomicTypes.stream().noneMatch(type -> type.sameType(dt))
        && !(dt instanceof DecimalType)) {
```

Let me see if `DataTypes.createDecimalType()` will work and ignore the precision / scale so they can all be kept in one set.

Reply:
No, `DataTypes.createDecimalType()` doesn't work as the type generated from that will compare with `precision` and `scale` to the input type to be truncated.

I think we can just remove this function, given that the actual call to bind that this function is used in does type validation already when choosing a specific input type (as follows):
```java
    DataType toTruncateDataType = toTruncateField.dataType();
    if (toTruncateDataType instanceof ByteType) {
      return new TruncateTinyInt();
    } else if (toTruncateDataType instanceof ShortType) {
      return new TruncateSmallInt();
    } else if (toTruncateDataType instanceof IntegerType) {
      return new TruncateInt();
    } else if (toTruncateDataType instanceof LongType) {
      return new TruncateBigInt();
    } else if (toTruncateDataType instanceof DecimalType) {
      return new TruncateDecimal(
          ((DecimalType) toTruncateDataType).precision(),
          ((DecimalType) toTruncateDataType).scale());
    } else if (toTruncateDataType instanceof StringType
        || toTruncateDataType instanceof VarcharType
        || toTruncateDataType instanceof CharType) {
      return new TruncateString();
    } else if (toTruncateDataType instanceof BinaryType) {
      return new TruncateBinary();
    } else {
      throw new UnsupportedOperationException("Cannot truncate type: " + toTruncateDataType);
    }
```

Reply:
I guess just the name is misleading. I wrote this before I saw how it was used. This isn't a complete list of truncateable types, it's just the ones that can be checked using `sameType`.

A better name would help. Also, `static final` constants should use names that are ALL_UPPER_CASE.

Reply:
We might need to normalize the Varchar / Char types, but otherwise I'll get rid of the function and just leave what is here.

Reply:
Got rid of the function and am now depending on this input block.